src/sinks/pulsar.rs

```diff
@@ -0,0 +1,178 @@
+        let producer = pulsar.producer(Some(ProducerOptions {
+            batch_size: config.batch_size,
+            ..ProducerOptions::default()
+        }));
+
+        Ok(Self {
+            topic: config.topic,
+            encoding: config.encoding,
+            pulsar,
+            producer,
+            in_flight: FuturesUnordered::new(),
+            seq_head: 0,
+            seq_tail: 0,
+            pending_acks: HashSet::new(),
+            acker,
+        })
+    }
+}
+
+impl Sink for PulsarSink {
+    type SinkItem = Event;
+    type SinkError = ();
+
+    fn start_send(&mut self, item: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
+        let message = encode_event(item, &self.topic, self.encoding).map_err(|_| ())?;
+        let fut = self.producer.send(self.topic.clone(), &PulsarSend(message));
+
+        let seqno = self.seq_head;
+        self.seq_head += 1;
+        self.in_flight
+            .push((Box::new(fut) as SendFuture).join(future::ok(seqno)));
+        Ok(AsyncSink::Ready)
+    }
+
+    fn poll_complete(&mut self) -> Poll<(), Self::SinkError> {
+        loop {
+            match self.in_flight.poll() {
+                Ok(Async::NotReady) => return Ok(Async::NotReady),
+                Ok(Async::Ready(None)) => return Ok(Async::Ready(())),
+                Ok(Async::Ready(Some((result, seqno)))) => {
+                    trace!(
+                        "produced message {:?} from {} at sequence id {}",
+                        result.message_id,
+                        result.producer_id,
+                        result.sequence_id
+                    );
+                    self.pending_acks.insert(seqno);
+                    let mut num_to_ack = 0;
+                    while self.pending_acks.remove(&self.seq_tail) {
+                        num_to_ack += 1;
+                        self.seq_tail += 1;
+                    }
+                    self.acker.ack(num_to_ack);
+                }
+                Err(e) => error!("future cancelled: {}", e),
+            }
+        }
+    }
+}
+
+// TODO: https://github.com/wyyerd/pulsar-rs/issues/60
+struct PulsarSend(Vec<u8>);
+
+impl SerializeMessage for PulsarSend {
+    fn serialize_message(input: &Self) -> Result<pulsar::producer::Message, pulsar::Error> {
+        Ok(pulsar::producer::Message {
+            payload: input.0.clone(),
+            ..Default::default()
+        })
+    }
+}
+
+fn encode_event<S: AsRef<str>>(item: Event, topic: S, enc: Encoding) -> crate::Result<Vec<u8>> {
+    let log = item.into_log();
+    let data = match enc {
+        Encoding::Json => serde_json::to_vec(&log.unflatten())?,
+        Encoding::Text => log
+            .get(&event::MESSAGE)
+            .map(|v| v.as_bytes().to_vec())
+            .unwrap_or_default(),
+    };
+    Ok(data)
+}
+
+fn healthcheck(config: &PulsarSinkConfig, pulsar: &Pulsar) -> crate::Result<super::Healthcheck> {
+    let consumer = pulsar
+        .consumer()
```

Comment:
I don't know in detail how Pulsar handles authorization, but trying to access the remote as a consumer in `healthcheck` but as a producer in the actual sink could hit different authorization paths.

I see the `kafka` sink also uses this approach, but it is a potential issue to be aware of.